hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/CloseableProcessorSupplier.java

```diff
@@ -23,45 +23,56 @@
 import javax.annotation.Nonnull;
 import java.io.Closeable;
 import java.util.Collection;
+import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
 import static com.hazelcast.jet.impl.util.ExceptionUtil.sneakyThrow;
-import static java.util.stream.Collectors.toList;
 
 /**
- * A {@link ProcessorSupplier} which closes created processor instances when
- * the job is complete. Class is useful to close external resources such as
- * connections or files.
+ * A {@link ProcessorSupplier} which closes the processor instances it
+ * created when the job is complete. Useful when the processor uses
+ * external resources such as connections or files.
  * <p>
- * The {@code close()} method is called after all other calls on any local
- * processor have returned.
+ * The processors must implement {@link Closeable}. The call to {@code
+ * close()} will be the very last call on the processor.
  *
- * @param <E> Processor type
+ * @param <E> the processor type
  */
 public class CloseableProcessorSupplier<E extends Processor & Closeable> implements ProcessorSupplier {
```

Comment:
we should mark the class as final

Reply:
we can't - we need to override the `init` method

Reply:
I don't think this class should be extensible. I would prefer that instead we take a Function<Context, Supplier<Processor>.

Reply:
as per discussion, let's just make the class abstract, create factory methods and add first-class `Processor.close()` support in future PR